23.0.0.6-beta staging (#3114)
OpenLiberty/blogs: posts/2023-06-13-23.0.0.6-beta.adoc

:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

Open Liberty 23.0.0.6-beta includes support for Private Key JWT authentication and Proof Key for Code Exchange (PKCE) for OpenID Connect clients.

This release also includes the latest updates to our ongoing link:{url-prefix}/blog/2023/02/10/instant-on-beta-update.html[InstantOn functionality], which provides incredibly fast startup times, as well as the latest link:{url-prefix}/blog/2023/04/18/23.0.0.4-beta.html[Jakarta Data preview]. 

If you'd like to know what else the team is currently considering or actively working on, refer to the https://github.com/orgs/OpenLiberty/projects/2[Open Liberty Roadmap] and the https://github.com/OpenLiberty/open-liberty/issues?q=is%3Aopen+is%3Aissue+label%3A%22In+Progress%22[in progress work items].

The link:{url-about}[Open Liberty] 23.0.0.6-beta includes the following beta features (along with link:{url-prefix}/docs/latest/reference/feature/feature-overview.html[all GA features]):

* <<pkjwt, Private Key JWT authentication support for OpenID Connect clients>>
* <<pkce, PKCE support for OpenID Connect clients>>

> See also link:{url-prefix}/blog/?search=beta&key=tag[previous Open Liberty beta blog posts].

Reviewer: ```suggestion
You can also check out our link:{url-prefix}/blog/?search=beta&key=tag[previous Open Liberty beta blog posts].
```

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/25333
// Contact/Reviewer: ayoho
// // // // // // // // 
[#pkjwt]
== Private Key JWT authentication support for OpenID Connect clients
OpenID Connect clients in Liberty now support the `private_key_jwt` client authentication method with OpenID Connect token endpoints. 

> OpenID Connect clients are required to provide authentication data to the OpenID Connect provider when invoking the provider's token endpoint. Clients can authenticate using several different methods, but most of those methods require a client secret. The new `private_key_jwt` authentication method allows clients to make use of asymmetric keys to create signed JSON Web Tokens (JWTs) to authenticate instead of client secrets. OpenID Connect clients using this authentication method are no longer required to have a client secret.

Reviewer: ```suggestion
OpenID Connect clients are required to provide authentication data to the OpenID Connect provider when they invoke the provider's token endpoint. Clients can authenticate using several different methods, but most of those methods require a client secret. The new `private_key_jwt` authentication method enables clients to use asymmetric keys to create signed JSON Web Tokens (JWTs) to authenticate instead of client secrets. OpenID Connect clients using this authentication method are no longer required to have a client secret.
```

> Server administrators can enable this functionality using the new `private_key_jwt` option for the `tokenEndpointAuthMethod` attribute, as well as the new `tokenEndpointAuthSigningAlgorithm` and `keyAliasName` attributes in either the `<openidConnectClient>` or `<oidcLogin>` elements in their `server.xml`.

Reviewer: ```suggestion
Server administrators can enable this functionality using the new `private_key_jwt` option for the `tokenEndpointAuthMethod` attribute, as well as the new `tokenEndpointAuthSigningAlgorithm` and `keyAliasName` attributes in either the `openidConnectClient` or `oidcLogin` elements in their `server.xml` file.
```

> For example, when using OpenID Connect Client feature, include:

Reviewer: ```suggestion
For example, when you use the OpenID Connect Client feature, include configuration similar to the following example:
```

[source,xml]
----
    <featureManager>
        <feature>openidConnectClient-1.0</feature>
    </featureManager>
    ...
    <openidConnectClient tokenEndpointAuthMethod="private_key_jwt" keyAliasName="privatekeyaliasRS512" ... />
----

> For example, when using Social Media Login feature, include:

Reviewer: ```suggestion
 If you are using the Social Media Login feature, include configuration similar to the following example:
```

[source,xml]
----
    <featureManager>
        <feature>socialLogin-1.0</feature>
    </featureManager>
    ...
    <oidcLogin tokenEndpointAuthMethod="private_key_jwt" tokenEndpointAuthSigningAlgorithm="E512" keyAliasName="privatekeyaliasES512" ... />
----

> The `tokenEndpointAuthSigningAlgorithm` attribute specifies the signing algorithm to use to sign the JWT used for client authentication. The `keyAliasName` attribute points to the key to use to sign the JWT.

Reviewer: ```suggestion
The `tokenEndpointAuthSigningAlgorithm` attribute specifies the signing algorithm to sign the JWT that is used for client authentication. The `keyAliasName` attribute points to the key to use to sign the JWT.
```

> More information about the configuration options can be found in the docs for the link:{url-prefix}/docs/latest/reference/config/openidConnectClient.html[openidConnectClient element] and the link:{url-prefix}/docs/latest/reference/config/oidcLogin.html[oidcLogin element]. 

Reviewer: ```suggestion
For more information about the configuration options, refer to the docs for the link:{url-prefix}/docs/latest/reference/config/openidConnectClient.html[openidConnectClient element] and the link:{url-prefix}/docs/latest/reference/config/oidcLogin.html[oidcLogin element]. 
```

> More information about `private_key_jwt` client authentication can be found in the link:https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication[OpenID Connect core specification] and link:https://datatracker.ietf.org/doc/html/rfc7523[RFC 7523].

Reviewer: ```suggestion
For more information about `private_key_jwt` client authentication, refer to the link:https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication[OpenID Connect core specification] and link:https://datatracker.ietf.org/doc/html/rfc7523[RFC 7523].
```

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC> 

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/25332
// Contact/Reviewer: ayoho
// // // // // // // // 
[#pkce]
== PKCE support for OpenID Connect clients
OpenID Connect clients in Open Liberty now support Proof Key for Code Exchange (PKCE) (RFC 7636). PKCE is an extension of the OAuth 2.0 specification that provides protection from authorization code interception attacks for OAuth 2.0 public clients. In very specific scenarios, a malicious application can intercept an authorization code intended for a legitimate OAuth 2.0 public client and use the authorization code to obtain access and ID tokens on behalf of the client. PKCE introduces additional steps and request parameters to prevent such interception attacks.

Server administrators can enable this functionality using the `pkceCodeChallengeMethod` attribute in either the `<openidConnectClient>` or `<oidcLogin>` elements in their `server.xml`.

> For example, when using OpenID Connect Client feature, include:

Reviewer: ```suggestion
For example, when you use the OpenID Connect Client feature, include configuration similar to the following example:
```

[source,xml]
----
    <featureManager>
        <feature>openidConnectClient-1.0</feature>
    </featureManager>
    ...
    <openidConnectClient pkceCodeChallengeMethod="S256" ... />
----

> For example, when using Social Media Login feature, include:

Reviewer: ```suggestion
 If you are using the Social Media Login feature, include configuration similar to the following example:
```

[source,xml]
----
    <featureManager>
        <feature>socialLogin-1.0</feature>
    </featureManager>
    ...
    <oidcLogin pkceCodeChallengeMethod="S256" ... />
----

> More information about the Liberty configuration options can be found in the docs for the link:{url-prefix}/docs/latest/reference/config/openidConnectClient.html[openidConnectClient element] and the link:{url-prefix}/docs/latest/reference/config/oidcLogin.html[oidcLogin element]. 

Reviewer: ```suggestion
For more information about the configuration options, refer to the docs for the link:{url-prefix}/docs/latest/reference/config/openidConnectClient.html[openidConnectClient element] and the link:{url-prefix}/docs/latest/reference/config/oidcLogin.html[oidcLogin element]. . 
```

> More information about PKCE can be found in link:https://datatracker.ietf.org/doc/html/rfc7636[RFC 7636].

Reviewer: ```suggestion
For more information about PKCE,  refer to link:https://datatracker.ietf.org/doc/html/rfc7636[RFC 7636].
```
